Add `RadioCard` component

alex-ju opened the pull request:
### :pushpin: Summary

Add `Hds::Form::RadioCard` component based on [current design](https://www.figma.com/file/noyY6dUMDYjmySpHcMjhkN/HDS-Product---Components?node-id=16108%3A37941).

[👉 Preview RadioCard component](https://hds-components-git-alex-ju-add-radio-card-component-hashicorp.vercel.app/components/form/radio-card)

### :hammer_and_wrench: Detailed description

We're introducing a `RadioCard` component comprising of:
 - `Hds::Form::RadioCard` (a bespoke component wrapping a `Form::Radio::Base` instance)
 - `Hds::Form::RadioCard::Group` (very much similar to its cousin `Hds::Form::Radio::Group`)

**Notes**
One interesting challenge with this component is the logic behind selecting a card (and deselecting the other cards), which is currently handled at the application level. After exploring a few options for internalizing this logic as part of the component I ended up leaving it as is, meaning an easier adoption process on the consumer side.

~~I am currently in the process of testing this component in `cloud-ui`, hence the draft status, but feel free to provide feedback in the meantime.~~ Tested this API in HCP Vault and HCP Consul (the main products using this component according to the audit).

I would like to add more examples under the 'How to use' section, showcasing a custom example (using the Generic block) and the `controlPosition` and `align` arguments.

### :link: External links

[Jira issue](https://hashicorp.atlassian.net/browse/HDS-613)

***

### 👀 How to review

<!-- Suggest how it's best for the reviewer to review the code (choose one, or remove) -->
👉 Review commit-by-commit
👉 Review by files changed

Reviewer's checklist:

- [x] +1 Percy if applicable
- [x] Confirm that PR has a changelog update via [Changesets](https://github.com/changesets/changesets) if needed

:speech_balloon: Please consider using [conventional comments](https://conventionalcomments.org/) when reviewing this PR.

didoo commented on packages/components/app/styles/components/form/radio-card.scss:
```diff
@@ -0,0 +1,131 @@
+
+.hds-form-group__radio-cards {
+  .hds-form-group__control-fields-wrapper {
+    gap: 16px;
```
we can't use `gap` yet because of the browsers versions that we need to support (see thread here: https://hashicorp.slack.com/archives/C11JCBJTW/p1635973292134000)

_(I've asked in Slack again, but I don't expect much difference from one year ago)_

KristinLBradley replied:
Just wanted to comment... From the Slack conversations it looks like John Cowen would be pretty open to updating the browsers they support so we could use flex gap.

https://hashicorp.slack.com/archives/C11JCBJTW/p1636022796147300?thread_ts=1635973292.134000&cid=C11JCBJTW
_"... Consul agreed to use a ~2 years back support, and thats what I'll be PRing today (hopefully). **But that is not set in stone at all.** It sounds like you have a straightforwards workaround, but **if it becomes too much of a workaround please reach out**"_

alex-ju replied:
I forgot about this open discussion. Until we clarify whether or not flex `gap` is something we're confident using across the products I switched to a `margin`-based approach.

didoo commented on packages/components/app/styles/components/form/radio-card.scss:
```diff
@@ -0,0 +1,131 @@
+// RadioCard Group
+
+.hds-form-group__radio-cards {
+  .hds-form-group__control-fields-wrapper {
```
do you think we could re-use the same approach we have for the "normal" control groups and add a "contextual" class `hds-form-group__control-field` to the radio-card items? or is too much complexity for no real value?

alex-ju replied:
I played a bit with this and the way this class is applied to a field element within a `fieldset` doesn't give us a clean way to pass a `contextualClass` to it as we're using the bare `yield` utility template for these fields. As an alternative, we can potentially do it at the `<Hds::Form::Fieldset>` level, passing it as an argument there, then using it when calling the field (e.g. <Hds::Form::Fieldset contextualFieldClass="hds-form-group__control-fields-wrapper">). I've not yet pushed any commits to address this, as we'll still have a few other overrides within `hds-form-group__radio-cards`, but if the goal is to get rid of these altogether, I can try some refactoring.

didoo commented on packages/components/tests/dummy/app/templates/components/form/radio-card.hbs:
```diff
@@ -0,0 +1,625 @@
+    <code class="code-dummy">RadioCard</code>
+    component:
+  </p>
+  <dl class="dummy-component-props" aria-labelledby="component-api-form-radio-card">
```
I _think_ from this APIs the following arguments/props are missing:
- `@controlPosition`
- `@align`

didoo commented on packages/components/tests/dummy/app/templates/components/form/radio-card.hbs:
```diff
@@ -0,0 +1,625 @@
+    <dt>&lt;[R].Generic&gt; <code>yielded component</code></dt>
+    <dd>
+      <p>It is a container that yields its content inside the card.</p>
+      <p>The content does not inherit any styles and can be customized as desired.</p>
```
Should we mention that, no matter what the order in the declaration in code, the `generic` content appears always at the bottom?

didoo replied:
was the thumbs up something todo, or just an ACK?

alex-ju replied:
I tried capturing [the rigid order of arguments (including `Generic` here](https://github.com/hashicorp/design-system/blob/alex-ju/add-radio-card-component/packages/components/tests/dummy/app/templates/components/form/radio-card.hbs#L104-L113)

didoo commented on packages/components/tests/dummy/app/templates/components/form/radio-card.hbs:
```diff
@@ -0,0 +1,625 @@
+        </em></p>
+    </dd>
+  </dl>
+  <h5 class="dummy-h5">Contextual components</h5>
```
as mentioned above, most of these contextual sub-component accept attributes spreading, so maybe we should make it explicit in the documentation

didoo commented on packages/components/tests/dummy/app/templates/components/form/radio-card.hbs:
```diff
@@ -0,0 +1,625 @@
+  </div>
+
+  <h4 class="dummy-h4">Card content</h4>
+  <Hds::Form::RadioCard::Group @name="radio-card-default" as |G|>
+    <G.Legend>Group legend</G.Legend>
+    {{#each @model.RADIOCARDS as |item|}}
+      <G.RadioCard @checked={{item.checked}} @value={{item.value}} {{on "change" this.onChange}} as |R|>
+        <R.Icon @name="hexagon" />
+        <R.Label>{{item.label}}</R.Label>
+        <R.Badge @text={{item.badge}} />
+        <R.Description>{{item.description}}</R.Description>
+      </G.RadioCard>
+    {{/each}}
+  </Hds::Form::RadioCard::Group>
```
Maybe in this section I would do something that stress-test the component in relation to its content: eg. one card with a longer label that spans multiple lines, one card with a multi-line description and one card without badge or icon. This would also allow a visual-check that all the cards have the same height even when the content is so different.

didoo replied:
was the thumbs up something todo, or just an ACK? on this one I feel more strongly, because it would allow us to test visually (also via visual regression testing) the working of the component

alex-ju replied:
I tried to vary the examples used across sections (including different heights) as follows:
How to use:
 - basic example (using the structured subcomponents Icon, Label, Badge, Description with varied lengths on the Label and Description)
 - custom example (using custom width, generic content, and multiple Badges).

Showcase:
 - contains examples with and without badges; can add one without icon – this may be the only one missing from the suggestions above.

didoo replied:
The issue I have with the "How to use" is that it's not covered by Percy. I would feel more "covered" and at peace of mind if the "variable height/content" example is in the showcase too (only one of the "rows", if you want).

Probably something like this :) :
```
  <span class="dummy-text-small">With different content height</span>
  <Hds::Form::RadioCard::Group @name="radio-card-group-custom" as |G|>
    <G.Legend>Group legend</G.Legend>
    <G.RadioCard @checked="true" {{on "change" this.onChange}} as |R|>
      <R.Icon @name="hexagon" />
      <R.Label>Radio card with normal label</R.Label>
      <R.Badge @text="Badge" />
      <R.Description>This is the radio card description text.</R.Description>
    </G.RadioCard>
    <G.RadioCard {{on "change" this.onChange}} as |R|>
      <R.Icon @name="hexagon" />
      <R.Label>Radio card with long label that spans multiple lines</R.Label>
      <R.Badge @text="Badge" />
      <R.Description>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Pellentesque erat elit, lacinia at magna
        eget, porttitor lobortis nulla.</R.Description>
    </G.RadioCard>
    <G.RadioCard {{on "change" this.onChange}} as |R|>
      <R.Label>Radio card without icon</R.Label>
      <R.Description>This is the radio card description text.</R.Description>
    </G.RadioCard>
  </Hds::Form::RadioCard::Group>
  ```

alex-ju replied:
Valid point about the 'How to use' section – have added the one above, and yes totally agree this example is a good way to prevent layout regression.

didoo reviewed (commented):
I've left a bunch of comments, but overall is very good and almost ready to go in production.

One thing I want to chat with you about is the _"logic behind selecting a card (and deselecting the other cards)"_, I want to know more about the issues you've found (for my knowledge).

alex-ju commented on packages/components/app/styles/components/form/radio-card.scss:
```diff
@@ -0,0 +1,131 @@
+
+  // ALIGNMENT
+
+  &--align-left {
```
Yes, I strongly agree with changing these in favor of findability – not sure why I did this in the first place 😅

KristinLBradley commented on packages/components/app/styles/components/form/radio-card.scss:
```diff
@@ -0,0 +1,131 @@
+
+  &:first-child {
+    margin-top: 0;
+  }
```
Nit: if a bottom margin were instead added to icons it seems you could remove this conditional style.

alex-ju replied:
yes, that was my initial intent, but `(component "flight-icon" size="24")` doesn't seem to allow for a ` class="hds-form-radio-card__icon"` to be passed in – I can dig a bit to understand why

Dhaulagiri replied:
I'm pretty rusty but it's quite possible you can't pass splattributes through component invocations of this type

alex-ju replied:
yep, you're right Brian, unless we do some argument forwarding (for example adding a `contextualClass` to the FlightIcon then setting `contextualClass="hds-form-radio-card__icon"` on the above call)

didoo replied:
yes, `class` can't be forwarded in contextual components (that's why we had to introduce `contextualClass` in other components)

probably I would not add this complexity and rely on the CSS as implemented now to solve the problem

KristinLBradley commented on packages/components/addon/components/hds/form/radio-card/index.hbs:
```diff
@@ -0,0 +1,20 @@
+      aria-describedby={{this.ariaDescribedBy}}
+      ...attributes
+    />
+  </div>
```
Nit: Technically the label element should only contain phrasing content (inline-level content) so should not be used to contain div tags.

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/label#technical_summary

alex-ju replied:
Yep 😬, aware of the spec, and while I'm not a supporter of `label`s wrapping `input`s I think in this very instance following [Melanie's prototype](https://codepen.io/melsumner/pen/oNEGJaw?editors=1100) has many benefits and reduces the complexity of this component (in terms of keeping element states in sync, etc).

KristinLBradley replied:
Hmm? It's fine for labels to wrap inputs. Just saying preferably we wouldn't use labels to wrap div tags since that's invalid. We could use span tags instead to make the markup validate.

KristinLBradley replied:
Cool, thanks! 

(CSS visual styling doesn't matter in terms of whether HTML markup is valid or not. :) )

KristinLBradley reviewed (commented):
See: https://hds-components-git-alex-ju-add-radio-card-component-hashicorp.vercel.app/components/form/radio-card#showcase

Question: Should the disabled state still have a hover effect? To me, when an element changes on hover it indicates that it's somehow interactive so it could possibly be confusing to users.

(_Clarification: I would personally not include any kind of visible hover effect on disabled elements since users may think it's interactive when it isn't._)

@heatherlarsen Opinions?

heatherlarsen commented:
@KristinLBradley I tend to agree with you, but this is a decision for @cveigt :)

cveigt commented:
@KristinLBradley I'm not opposed to it. I think it's a good point; will explore the option in Figma.

cveigt commented on packages/components/tests/dummy/app/templates/components/form/radio-card.hbs:
```diff
@@ -0,0 +1,625 @@
+      </p>
+      <p>Default: <span class="default">false</span></p>
+    </dd>
+    <dt>isOptional <code>boolean</code></dt>
```
We already discussed it; just a note about removing _isOptional_ from the page.

cveigt reviewed (changes requested):
#### Legend
One difference from the radio card group and radio groups in that the spacing between the legend and the card will be 12px and not 8px.
![Screen Shot 2022-09-21 at 11 54 49 PM](https://user-images.githubusercontent.com/2142312/191678607-18c7972e-dbd2-4409-9370-898166337b41.png)

#### Group width
Question: How can the group's width be managed/controlled? Like fixed vs full width? 👇
![Screen Shot 2022-09-22 at 12 07 43 AM](https://user-images.githubusercontent.com/2142312/191680977-605f5464-57ac-487b-8c96-a481935fe1a5.png)

#### Alignment vs Align
Even though the flexbox property is called align, alignment feels clearer to me. I also looked at @jorytindall’s table and the table header and table cell have an alignment property, so I’d like to be consistent with the naming when possible, though I couldn’t find the same property in the code API for the table; I wonder if it’s called the same or not.

#### Disabled attribute
I followed @KristinLBradley's suggestion and removed any elevation from the disabled card, leaving only the border. I opened a new [Figma branch](https://www.figma.com/file/noyY6dUMDYjmySpHcMjhkN/branch/5qLTIwHeJl66yJAq3BAV0u/HDS-Product---Components?node-id=16108%3A37941) for these updates.

#### Motion
I want to explore an animation to make the transitions between states a bit smother.

jorytindall commented:
@cveigt I'm not sure that the `alignment` property has been implemented in the draft table component yet. I would imagine that they would both be `alignment` in Figma and in the component code though.

didoo commented:
> @cveigt I'm not sure that the `alignment` property has been implemented in the draft table component yet. I would imagine that they would both be `alignment` in Figma and in the component code though.

from a developer perspective `align` works better for me than `alignment` (reason: the word commonly used in CSS is `align` not `alignment`). in any case I don't see an issue if is `align` in code and `alignment` in Figma, they don't generate confusion, it's clear they're the same

alex-ju commented on packages/components/app/styles/components/form/radio-card.scss:
```diff
@@ -0,0 +1,131 @@
+  display: flex;
+  flex: 1;
+  flex-direction: column;
+  font-family: var(--token-typography-font-stack-text);
```
old habits – I was used to avoiding class composition, hence it didn't cross my mind to use the typography classes (with which I need to get more accustomed); removed it from here and used classes on relevant elements.

KristinLBradley commented:
> 

@cveigt
For the question of managing the width of RadioCards...

We could set a max-width so that the width wouldn't expand past a certain value but would squish down as needed within layouts. For determining the max-width to use, I would look at the most likely use cases within layouts. IE: a single card in a row vs. multiple side by side cards such as 2 or 3 in a row.

cveigt commented:
> from a developer perspective align works better for me than alignment (reason: the word commonly used in CSS is align not alignment). in any case I don't see an issue if is align in code and alignment in Figma, they don't generate confusion, it's clear they're the same

Ok, sounds good. I'll leave the decision of how it'll be called in code up to you (eng). I'll keep it as `alignment` in Figma.

cveigt commented:
@alex-ju Regarding the transition, I landed on `0.2s`. Here's a [codepen](https://codepen.io/cveigt/pen/BaxwogQ) for reference. Let me know what you think 😬

alex-ju commented:
Thanks for the thorough review @cveigt! I addressed all your comments in the recent update.

> #### Legend
> One difference from the radio card group and radio groups in that the spacing between the legend and the card will be 

Good spot – have updated through an override for now. I'll check to see if there is scope in passing a modifier class to cope with this difference.

> #### Group width
> Question: How can the group's width be managed/controlled? Like fixed vs full width? 👇

There are a few options here:
 - one is to leave the with constrain to the consumer
 - another is for us to expose a `@maxWidth` attribute (similarly to how we expose `@width` in other components and to what Kristin mentioned above), allowing consumers to constrain the cards without needing to declare a CSS property.
I am leaning more towards the second option, so I will add the attribute and expose an example in the 'Showcase' section.

> #### Alignment vs Align
> Even though the flexbox property is called align, alignment feels clearer to me. I also looked at @jorytindall’s table and the table header and table cell have an alignment property, so I’d like to be consistent with the naming when possible, though I couldn’t find the same property in the code API for the table; I wonder if it’s called the same or not.

As discussed, I don't have strong opinions on this, just acknowledged the different notation in the CSS world. I have updated the component argument to `alignment` to keep it in line with Figma and the table component.

> #### Disabled attribute
> I followed @KristinLBradley's suggestion and removed any elevation from the disabled card, leaving only the border. I opened a new [Figma branch](https://www.figma.com/file/noyY6dUMDYjmySpHcMjhkN/branch/5qLTIwHeJl66yJAq3BAV0u/HDS-Product---Components?node-id=16108%3A37941) for these updates.

Yep, great suggestion here, have updated.

> #### Motion
> I want to explore an animation to make the transitions between states a bit smother.
> Update: Regarding the transition, I landed on 0.2s. Here's a [codepen](https://codepen.io/cveigt/pen/BaxwogQ) for reference. Let me know what you think

Have added the transition – it feels/looks good to me

KristinLBradley commented on packages/components/app/styles/components/form/radio-card.scss:
```diff
@@ -0,0 +1,131 @@
+
+.hds-form-radio-card {
+  display: flex;
+  flex: 1;
```
Personally, I wouldn't set this here as a hard-coded style (`flex: 1;`). I would let the layout be determined by a parent wrapper element of the end-user's choice.

alex-ju replied:
@didoo agree that the `<Hds::Form::RadioCard>` shouldn't make assumptions about the parent layout – one option is to only apply this attribute when the `<Hds::Form::RadioCard>` is under a `<Hds::Form::RadioCard::Group>`.

@KristinLBradley the most common scenario is where a set of `<Hds::Form::RadioCard>` components are wrapped with a `<Hds::Form::RadioCard::Group>`. In this case, how would the consumer wrap the `<Hds::Form::RadioCard>`s? Would you be able to exemplify the above suggestion in code? I feel that I'm missing something here.

KristinLBradley replied:
I think your proposal of applying it when `<Hds::Form::RadioCard>` is within a `<Hds::Form::RadioCard::Group>` makes sense.

KristinLBradley commented:
> 

I would vote for using "align". To me it's clearer (and also shorter which is nice).

alex-ju commented on packages/components/app/styles/components/form/radio-card.scss:
```diff
@@ -0,0 +1,131 @@
+  font-family: var(--token-typography-font-stack-text);
+  background-color: var(--token-color-surface-primary);
+  border: 1px solid var(--token-color-border-primary);
+  border-radius: 6px;
```
Sure, sounds like a good idea and can do in a follow-up PR.

didoo commented on packages/components/tests/integration/components/hds/form/radio-card/index-test.js:
```diff
@@ -0,0 +1,84 @@
+    assert.dom('input').hasAttribute('data-test1');
+    assert.dom('input').hasAttribute('data-test2', 'test');
+  });
+});
```
was the thumbs up something todo, or just an ACK?

alex-ju replied:
Thumbs up were meant to be acked and addressed 😅  (nothing scientific though) – I [added tests for thrown assertions here](https://github.com/hashicorp/design-system/blob/alex-ju/add-radio-card-component/packages/components/tests/integration/components/hds/form/radio-card/index-test.js#L31-L57)

didoo replied:
@alex-ju super nitpick: assertions are always declared at the bottom of the tests :|

alex-ju replied:
Updated

alex-ju commented:
> OK, all clear. I _think_ what I would do in this case is to add a "notice" block in the documentation, in the "how to use" section, which explain this peculiarity, so that whoever needs to implement radio cards, they know they have to manage themselves the state logic.

Added a note in the ['How to use' section](http://localhost:4200/components/form/radio-card#how-to-use), let me know if it sounds clear enough.

didoo reviewed (approved):
All good for me! 🛳 !

cveigt reviewed (approved):
Looks great. Thanks for making that last change! 🚀

alex-ju commented:
Rebased to `main` – planning to merge this shortly after and follow up to add the 'Design guidelines' and potentially convert some of the CSS values into tokens.
